File: blender_extension/operations/camera.py
"""Camera operations, including automatic framing."""

from __future__ import annotations

import math
from typing import Any

import bpy
from mathutils import Vector

from .. import ids
from ..dispatcher import op, read
from ..protocol import BridgeError, ErrorCode, invalid_argument
from . import _common as c

TRACK_CONSTRAINTS = {"TRACK_TO", "DAMPED_TRACK", "LOCKED_TRACK"}


def require_camera(obj):
    if obj.type != "CAMERA":
        raise BridgeError(
            ErrorCode.CAMERA_NOT_FOUND,
            f"`{obj.name}` is a {obj.type} object, not a camera.",
            {"object": obj.name, "type": obj.type},
        )
    return obj.data


def active_camera():
    camera = bpy.context.scene.camera
    if camera is None:
        raise BridgeError(
            ErrorCode.CAMERA_NOT_FOUND,
            "The scene has no active camera. Create one, or name a camera explicitly.",
            {"scene": bpy.context.scene.name},
        )
    return camera


def camera_arg(args: dict, key: str = "camera", required: bool = False):
    reference = c.optional_str(args, key)
    if reference is None:
        if required:
            raise invalid_argument(f"`{key}` is required.", field=key)
        return active_camera()
    obj = ids.find_object(reference)
    require_camera(obj)
    return obj


def summarise(obj, *, detail: bool = False) -> dict[str, Any]:
    camera = obj.data
    payload: dict[str, Any] = {
        "id": ids.ensure_id(obj),
        "data_id": ids.ensure_id(camera),
        "name": obj.name,
        "location": c.vector_dict(obj.location),
        "rotation_euler": c.vector_dict(obj.rotation_euler),
        "lens_mm": float(camera.lens),
        "sensor_width": float(camera.sensor_width),
        "sensor_height": float(camera.sensor_height),
        "projection": camera.type,
        "clip_start": float(camera.clip_start),
        "clip_end": float(camera.clip_end),
        "is_active": bpy.context.scene.camera == obj,
    }
    if camera.type == "ORTHO":
        payload["ortho_scale"] = float(camera.ortho_scale)
    dof = camera.dof
    payload["depth_of_field"] = {
        "enabled": bool(dof.use_dof),
        "focus_object": dof.focus_object.name if dof.focus_object else None,
        "focus_distance": float(dof.focus_distance),
        "f_stop": float(dof.aperture_fstop),
        "blades": int(dof.aperture_blades),
        "rotation": float(dof.aperture_rotation),
        "ratio": float(dof.aperture_ratio),
    }
    if obj.constraints:
        payload["constraints"] = [con.type for con in obj.constraints]
    if detail:
        payload["shift"] = {"x": float(camera.shift_x), "y": float(camera.shift_y)}
        payload["sensor_fit"] = camera.sensor_fit
        payload["fov_degrees"] = math.degrees(camera.angle)
    return payload


def apply_settings(obj, args: dict) -> list[str]:
    camera = obj.data
    changed: list[str] = []

    lens = c.optional(args, "lens")
    if lens is not None:
        if not isinstance(lens, dict) or len(lens) != 1:
            raise invalid_argument(
                "`lens` must be {\"millimetres\": n} or {\"fov_degrees\": n}.", field="lens"
            )
        kind, value = next(iter(lens.items()))
        if kind == "millimetres":
            camera.lens_unit = "MILLIMETERS"
            camera.lens = float(value)
        elif kind == "fov_degrees":
            camera.lens_unit = "FOV"
            camera.angle = math.radians(float(value))
        else:
            raise invalid_argument(
                "`lens` must be {\"millimetres\": n} or {\"fov_degrees\": n}.", field="lens"
            )
        changed.append("lens")

    projection = c.optional_str(args, "projection")
    if projection is not None:
        camera.type = {
            "PERSPECTIVE": "PERSP",
            "ORTHOGRAPHIC": "ORTHO",
            "PANORAMIC": "PANO",
        }[c.enum_value(projection, ["PERSPECTIVE", "ORTHOGRAPHIC", "PANORAMIC"], "projection")]
        changed.append("projection")

    for key, attribute in (
        ("ortho_scale", "ortho_scale"),
        ("sensor_width", "sensor_width"),
        ("sensor_height", "sensor_height"),
        ("clip_start", "clip_start"),
        ("clip_end", "clip_end"),
    ):
        value = c.optional_float(args, key)
        if value is not None:
            setattr(camera, attribute, value)
            changed.append(key)

    sensor_fit = c.optional_str(args, "sensor_fit")
    if sensor_fit is not None:
        camera.sensor_fit = c.enum_value(
            sensor_fit, ["AUTO", "HORIZONTAL", "VERTICAL"], "sensor_fit"
        )
        changed.append("sensor_fit")

    shift = c.optional(args, "shift")
    if shift is not None:
        camera.shift_x = float(shift.get("x", camera.shift_x))
        camera.shift_y = float(shift.get("y", camera.shift_y))
        changed.append("shift")

    dof_args = c.optional(args, "depth_of_field")
    if dof_args:
        changed.extend(_apply_dof(camera, dof_args))

    return changed


def _apply_dof(camera, args: dict) -> list[str]:
    dof = camera.dof
    changed: list[str] = []
    if args.get("enabled") is not None:
        dof.use_dof = bool(args["enabled"])
        changed.append("dof.enabled")
    if args.get("focus_object") is not None:
        dof.focus_object = ids.find_object(str(args["focus_object"]))
        dof.use_dof = True
        changed.append("dof.focus_object")
    if args.get("focus_distance") is not None:
        dof.focus_distance = float(args["focus_distance"])
        dof.use_dof = True
        changed.append("dof.focus_distance")
    if args.get("f_stop") is not None:
        dof.aperture_fstop = float(args["f_stop"])
        changed.append("dof.f_stop")
    if args.get("blades") is not None:
        dof.aperture_blades = int(args["blades"])
        changed.append("dof.blades")
    if args.get("rotation") is not None:
        dof.aperture_rotation = float(args["rotation"])
        changed.append("dof.rotation")
    if args.get("ratio") is not None:
        dof.aperture_ratio = float(args["ratio"])
        changed.append("dof.ratio")
    return changed


def aim(obj, point: Vector) -> None:
    """Point a camera at a world location. A camera looks down its local -Z."""
    direction = point - obj.matrix_world.translation
    if direction.length < 1e-9:
        return
    obj.rotation_mode = "XYZ"
    obj.rotation_euler = direction.to_track_quat("-Z", "Y").to_euler("XYZ")


def subject_bounds(objects: list) -> tuple[Vector, Vector]:
    bounds = c.world_bounds(objects)
    if bounds is None:
        raise invalid_argument("The subject has no geometry to frame.")
    return bounds


# --- operations ------------------------------------------------------------


@read("camera.list")
def list_cameras(ctx, args: dict) -> dict[str, Any]:
    name_filter = c.optional_str(args, "name_contains")
    matched = [
        obj
        for obj in bpy.data.objects
        if obj.type == "CAMERA" and c.matches_name(obj.name, name_filter)
    ]
    matched.sort(key=lambda o: o.name)
    window, cursor = c.paginate(matched, args)
    return {
        "cameras": [summarise(obj) for obj in window],
        "total": len(matched),
        "next_cursor": cursor,
        "revision": ctx.revision,
    }


@read("camera.get")
def get(ctx, args: dict) -> dict[str, Any]:
    obj = camera_arg(args)
    return {"camera": summarise(obj, detail=True), "revision": ctx.revision}


@op("camera.create")
def create(ctx, args: dict) -> dict[str, Any]:
    name = c.optional_str(args, "name") or "Camera"
    data = bpy.data.cameras.new(name)
    obj = bpy.data.objects.new(name, data)

    collection = c.collection_arg(args, "collection") or bpy.context.scene.collection
    collection.objects.link(obj)

    location = c.optional_vector(args, "location")
    if location is not None:
        obj.location = location

    apply_settings(obj, args)

    frame_targets = c.objects_arg(args, "frame_objects", required=False)
    if frame_targets:
        _frame(obj, frame_targets, args)
    else:
        rotation = c.optional_vector(args, "rotation")
        look_at = c.optional_vector(args, "look_at")
        if rotation is not None:
            obj.rotation_mode = "XYZ"
            obj.rotation_euler = rotation
        elif look_at is not None:
            bpy.context.view_layer.update()
            aim(obj, look_at)

    if c.optional_bool(args, "set_active", False):
        bpy.context.scene.camera = obj

    ids.invalidate_cache("object")
    ctx.bump()
    return {"camera": summarise(obj, detail=True), "revision": ctx.revision}


@op("camera.update")
def update(ctx, args: dict) -> dict[str, Any]:
    obj = camera_arg(args, required=True)
    changed: list[str] = []

    name = c.optional_str(args, "name")
    if name is not None:
        obj.name = name
        obj.data.name = name
        changed.append("name")
        ids.invalidate_cache("object")

    changed.extend(apply_settings(obj, args))
    if not changed:
        raise invalid_argument("Nothing to update on this camera.")

    ctx.bump()
    return {"camera": summarise(obj, detail=True), "changed": changed, "revision": ctx.revision}


@op("camera.delete")
def delete(ctx, args: dict) -> dict[str, Any]:
    obj = camera_arg(args, required=True)
    payload = {"id": ids.ensure_id(obj), "name": obj.name}
    data = obj.data
    bpy.data.objects.remove(obj, do_unlink=True)
    if data is not None and data.users == 0:
        bpy.data.cameras.remove(data)
    ids.invalidate_cache("object")
    ctx.bump()
    return {"deleted": payload, "revision": ctx.revision}


@op("camera.set_active")
def set_active(ctx, args: dict) -> dict[str, Any]:
    obj = camera_arg(args, required=True)
    bpy.context.scene.camera = obj
    ctx.bump()
    return {"active_camera": ids.ensure_id(obj), "name": obj.name, "revision": ctx.revision}


@op("camera.look_at")
def look_at(ctx, args: dict) -> dict[str, Any]:
    obj = camera_arg(args)
    point = c.optional_vector(args, "point")
    if point is None:
        target = c.optional_str(args, "target")
        if target is None:
            raise invalid_argument("Provide `point` or `target`.")
        minimum, maximum = subject_bounds([ids.find_object(target)])
        point = (minimum + maximum) * 0.5

    bpy.context.view_layer.update()
    aim(obj, point)
    ctx.bump()
    return {"camera": summarise(obj), "aimed_at": c.vector_dict(point), "revision": ctx.revision}


@op("camera.track_object")
def track_object(ctx, args: dict) -> dict[str, Any]:
    obj = camera_arg(args)
    target = c.object_arg(args, "target")
    constraint_type = c.enum_value(
        c.optional_str(args, "constraint", "TRACK_TO") or "TRACK_TO",
        sorted(TRACK_CONSTRAINTS),
        "constraint",
    )
    if obj == target:
        raise invalid_argument("A camera cannot track itself.")

    # Replace any existing tracking constraint rather than stacking a second
    # one, which would fight the first.
    for existing in [con for con in obj.constraints if con.type in TRACK_CONSTRAINTS]:
        obj.constraints.remove(existing)

    constraint = obj.constraints.new(type=constraint_type)
    constraint.target = target
    if constraint_type == "TRACK_TO":
        constraint.track_axis = "TRACK_NEGATIVE_Z"
        constraint.up_axis = "UP_Y"
    elif constraint_type == "DAMPED_TRACK":
        constraint.track_axis = "TRACK_NEGATIVE_Z"

    if c.optional_bool(args, "focus_on_target", False):
        obj.data.dof.use_dof = True
        obj.data.dof.focus_object = target

    ctx.bump()
    return {
        "camera": summarise(obj),
        "constraint": constraint_type,
        "target": ids.ensure_id(target),
        "revision": ctx.revision,
    }


@op("camera.clear_tracking")
def clear_tracking(ctx, args: dict) -> dict[str, Any]:
    obj = camera_arg(args)
    removed = []
    for existing in [con for con in obj.constraints if con.type in TRACK_CONSTRAINTS]:
        removed.append(existing.type)
        obj.constraints.remove(existing)
    ctx.bump()
    return {"camera": summarise(obj), "removed": removed, "revision": ctx.revision}


@op("camera.depth_of_field.update")
def update_dof(ctx, args: dict) -> dict[str, Any]:
    obj = camera_arg(args)
    changed = _apply_dof(obj.data, args)
    if not changed:
        raise invalid_argument("Nothing to change about the depth of field.")
    ctx.bump()
    return {"camera": summarise(obj, detail=True), "changed": changed, "revision": ctx.revision}


@op("camera.auto_frame")
def auto_frame(ctx, args: dict) -> dict[str, Any]:
    """Place and aim a camera so the subject fills the frame.

    The distance calculation lives here rather than being iterated towards:
    given the subject bounding sphere and the camera field of view, the
    distance that fits it is a closed-form expression.
    """
    obj = camera_arg(args)
    targets = c.objects_arg(args, "objects", required=False)
    if not targets:
        targets = [
            candidate
            for candidate in bpy.context.scene.objects
            if candidate.type in {"MESH", "CURVE", "FONT", "SURFACE", "META"}
            and candidate.visible_get()
        ]
    if not targets:
        raise invalid_argument("There is nothing visible to frame.")

    result = _frame(obj, targets, args)
    ctx.bump()
    return {"camera": summarise(obj), **result, "revision": ctx.revision}


def _frame(obj, targets: list, args: dict) -> dict[str, Any]:
    padding = c.optional_float(args, "padding", 0.1) or 0.0
    keep_position = c.optional_bool(args, "keep_position", False)
    should_aim = c.optional_bool(args, "aim", True)
    focus = c.optional_bool(args, "focus", False)
    direction_arg = c.optional_vector(args, "direction")

    bpy.context.view_layer.update()
    minimum, maximum = subject_bounds(targets)
    centre = (minimum + maximum) * 0.5
    radius = max((maximum - minimum).length * 0.5, 1e-6)

    camera = obj.data
    render = bpy.context.scene.render
    aspect = (render.resolution_x * render.pixel_aspect_x) / max(
        render.resolution_y * render.pixel_aspect_y, 1e-6
    )

    # Fit the bounding sphere in whichever of the two field-of-view angles is
    # tighter, so nothing is cropped in portrait or in landscape.
    horizontal_fov = camera.angle_x if hasattr(camera, "angle_x") else camera.angle
    vertical_fov = 2.0 * math.atan(math.tan(horizontal_fov * 0.5) / max(aspect, 1e-6))
    tightest = min(horizontal_fov, vertical_fov)
    distance = (radius * (1.0 + padding)) / max(math.sin(tightest * 0.5), 1e-6)

    if camera.type == "ORTHO":
        camera.ortho_scale = 2.0 * radius * (1.0 + padding)

    if not keep_position:
        if direction_arg is not None and direction_arg.length > 1e-9:
            direction = direction_arg.normalized()
        else:
            current = Vector(obj.matrix_world.translation) - centre
            if current.length > 1e-6:
                direction = current.normalized()
            else:
                # A three-quarter view from the front-left and slightly above:
                # the standard product-shot angle, and far more informative
                # than an axis-aligned one.
                direction = Vector((-0.6, -0.7, 0.38)).normalized()
        obj.location = centre + direction * distance

    bpy.context.view_layer.update()
    if should_aim:
        aim(obj, centre)

    if focus:
        camera.dof.use_dof = True
        camera.dof.focus_distance = (Vector(obj.matrix_world.translation) - centre).length

    return {
        "framed": [ids.ensure_id(target) for target in targets],
        "center": c.vector_dict(centre),
        "radius": radius,
        "distance": distance,
        "bounds": {"min": c.vector_dict(minimum), "max": c.vector_dict(maximum)},
    }

File: crates/blender-protocol/src/camera.rs
//! Camera payloads, including automatic framing.

use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

use crate::{
    BlenderError, Page, Result, Validate, check_name,
    ids::{CameraId, CollectionRef, ObjectRef},
    math::{Finite, Vec2, Vec3, check_positive, check_range},
};

/// Camera projection.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum ProjectionType {
    Perspective,
    Orthographic,
    Panoramic,
}

/// How the lens is specified.
#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize, JsonSchema)]
#[serde(rename_all = "snake_case")]
pub enum Lens {
    /// Focal length in millimetres.
    Millimetres(f64),
    /// Horizontal field of view in degrees.
    FovDegrees(f64),
}

impl Lens {
    /// Focal length for a given sensor width.
    pub fn focal_length(self, sensor_width_mm: f64) -> f64 {
        match self {
            Lens::Millimetres(mm) => mm,
            Lens::FovDegrees(degrees) => {
                let half = (degrees.to_radians() * 0.5).tan();
                if half.abs() < 1e-9 {
                    f64::MAX
                } else {
                    (sensor_width_mm * 0.5) / half
                }
            }
        }
    }

    /// Horizontal field of view in radians for a given sensor width.
    pub fn fov_radians(self, sensor_width_mm: f64) -> f64 {
        match self {
            Lens::Millimetres(mm) => 2.0 * ((sensor_width_mm * 0.5) / mm).atan(),
            Lens::FovDegrees(degrees) => degrees.to_radians(),
        }
    }
}

impl Validate for Lens {
    fn validate(&self) -> Result<()> {
        match self {
            Lens::Millimetres(mm) => check_range(*mm, 1.0, 5000.0, "lens.millimetres"),
            Lens::FovDegrees(degrees) => check_range(*degrees, 0.1, 179.0, "lens.fov_degrees"),
        }
    }
}

/// Depth-of-field settings.
#[derive(Debug, Clone, Default, Serialize, Deserialize, JsonSchema)]
pub struct DepthOfField {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub enabled: Option<bool>,
    /// Object to keep in focus. Overrides `focus_distance`.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub focus_object: Option<ObjectRef>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub focus_distance: Option<f64>,
    /// Aperture. Lower is shallower depth of field.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub f_stop: Option<f64>,
    /// Aperture blade count. 0 gives a perfectly circular bokeh.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub blades: Option<u32>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub rotation: Option<f64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub ratio: Option<f64>,
}

impl Validate for DepthOfField {
    fn validate(&self) -> Result<()> {
        if let Some(distance) = self.focus_distance {
            check_positive(distance, "focus_distance")?;
        }
        if let Some(f_stop) = self.f_stop {
            check_range(f_stop, 0.1, 128.0, "f_stop")?;
        }
        if let Some(blades) = self.blades
            && blades != 0
            && !(3..=16).contains(&blades)
        {
            return Err(BlenderError::invalid_argument(format!(
                "`blades` must be 0 (circular) or between 3 and 16, got {blades}."
            )));
        }
        if let Some(ratio) = self.ratio {
            check_positive(ratio, "ratio")?;
        }
        if self.focus_object.is_some() && self.focus_distance.is_some() {
            return Err(BlenderError::invalid_argument(
                "`focus_object` and `focus_distance` conflict; a focus object drives the distance.",
            ));
        }
        Ok(())
    }
}

/// Camera data-block settings.
#[derive(Debug, Clone, Default, Serialize, Deserialize, JsonSchema)]
pub struct CameraSettings {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub lens: Option<Lens>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub projection: Option<ProjectionType>,
    /// Orthographic view height in Blender units.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub ortho_scale: Option<f64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub sensor_width: Option<f64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub sensor_height: Option<f64>,
    /// `AUTO`, `HORIZONTAL` or `VERTICAL`.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub sensor_fit: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub clip_start: Option<f64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub clip_end: Option<f64>,
    /// Lens shift, in units of sensor width.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub shift: Option<Vec2>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub depth_of_field: Option<DepthOfField>,
}

impl Validate for CameraSettings {
    fn validate(&self) -> Result<()> {
        if let Some(lens) = self.lens {
            lens.validate()?;
        }
        for (value, field) in [
            (self.ortho_scale, "ortho_scale"),
            (self.sensor_width, "sensor_width"),
            (self.sensor_height, "sensor_height"),
            (self.clip_start, "clip_start"),
            (self.clip_end, "clip_end"),
        ] {
            if let Some(v) = value {
                check_positive(v, field)?;
            }
        }
        if let (Some(start), Some(end)) = (self.clip_start, self.clip_end)
            && start >= end
        {
            return Err(BlenderError::invalid_argument(format!(
                "`clip_start` ({start}) must be less than `clip_end` ({end})."
            )));
        }
        if let Some(shift) = self.shift {
            shift.check_finite("shift")?;
        }
        if let Some(fit) = &self.sensor_fit {
            const FITS: [&str; 3] = ["AUTO", "HORIZONTAL", "VERTICAL"];
            if !FITS.contains(&fit.as_str()) {
                return Err(BlenderError::invalid_enum("sensor_fit", fit.clone(), FITS));
            }
        }
        if let Some(dof) = &self.depth_of_field {
            dof.validate()?;
        }
        Ok(())
    }
}

/// `camera.create`
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct CreateCamera {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub location: Option<Vec3>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub rotation: Option<Vec3>,
    /// Aim at this point instead of specifying a rotation.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub look_at: Option<Vec3>,
    /// Frame these objects instead of specifying a location.
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub frame_objects: Vec<ObjectRef>,
    /// Make this the scene's active camera.
    #[serde(default)]
    pub set_active: bool,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub collection: Option<CollectionRef>,
    #[serde(default, flatten)]
    pub settings: CameraSettings,
}

impl Validate for CreateCamera {
    fn validate(&self) -> Result<()> {
        if let Some(name) = &self.name {
            check_name(name, "name")?;
        }
        self.location.check_finite("location")?;
        self.rotation.check_finite("rotation")?;
        self.look_at.check_finite("look_at")?;
        if self.rotation.is_some() && self.look_at.is_some() {
            return Err(BlenderError::invalid_argument(
                "Set `rotation` or `look_at`, not both.",
            ));
        }
        if !self.frame_objects.is_empty() && self.location.is_some() {
            return Err(BlenderError::invalid_argument(
                "`frame_objects` computes the location; remove `location` or remove `frame_objects`.",
            ));
        }
        self.settings.validate()
    }
}

/// `camera.update`
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct UpdateCamera {
    pub camera: ObjectRef,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    #[serde(default, flatten)]
    pub settings: CameraSettings,
}

impl Validate for UpdateCamera {
    fn validate(&self) -> Result<()> {
        if let Some(name) = &self.name {
            check_name(name, "name")?;
        }
        self.settings.validate()
    }
}

/// `camera.track_object` -- add a constraint so the camera keeps aiming at a
/// target as either moves.
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct TrackObject {
    pub camera: ObjectRef,
    pub target: ObjectRef,
    /// `TRACK_TO` (with an up axis) or `DAMPED_TRACK` (shortest rotation).
    #[serde(default = "default_track_type")]
    pub constraint: TrackConstraint,
    /// Also set the camera's depth of field focus to the target.
    #[serde(default)]
    pub focus_on_target: bool,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum TrackConstraint {
    TrackTo,
    DampedTrack,
    LockedTrack,
}

fn default_track_type() -> TrackConstraint {
    TrackConstraint::TrackTo
}

impl Validate for TrackObject {
    fn validate(&self) -> Result<()> {
        if self.camera == self.target {
            return Err(BlenderError::invalid_argument(
                "A camera cannot track itself.",
            ));
        }
        Ok(())
    }
}

/// `camera.auto_frame` -- place the camera so the given objects fill the frame.
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct AutoFrame {
    /// Camera to move. Omit to use the scene's active camera.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub camera: Option<ObjectRef>,
    /// Objects to fit in frame. Empty means every visible object.
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub objects: Vec<ObjectRef>,
    /// Extra margin around the subject, as a fraction of its size.
    #[serde(default = "default_padding")]
    pub padding: f64,
    /// Direction to view from, in world space. Normalised server-side.
    /// Defaults to a three-quarter view from the front-left and slightly above.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub direction: Option<Vec3>,
    /// Keep the camera where it is and only adjust the lens.
    #[serde(default)]
    pub keep_position: bool,
    /// Also aim the camera at the subject's centre.
    #[serde(default = "crate::object::default_true")]
    pub aim: bool,
    /// Set the depth-of-field focus distance to the subject's centre.
    #[serde(default)]
    pub focus: bool,
}

fn default_padding() -> f64 {
    0.1
}

impl Validate for AutoFrame {
    fn validate(&self) -> Result<()> {
        check_range(self.padding, 0.0, 10.0, "padding")?;
        if let Some(direction) = self.direction {
            direction.check_finite("direction")?;
            if direction.normalized().is_none() {
                return Err(BlenderError::invalid_argument(
                    "`direction` has zero length, so it does not describe a direction.",
                ));
            }
        }
        if self.keep_position && self.direction.is_some() {
            return Err(BlenderError::invalid_argument(
                "`direction` moves the camera, which `keep_position` forbids.",
            ));
        }
        Ok(())
    }
}

/// A camera as reported by the bridge.
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct CameraSummary {
    /// Id of the camera *object*.
    pub id: crate::ids::ObjectId,
    /// Id of the camera data-block.
    pub data_id: CameraId,
    pub name: String,
    pub location: Vec3,
    pub rotation_euler: Vec3,
    pub lens_mm: f64,
    pub sensor_width: f64,
    pub sensor_height: f64,
    pub projection: ProjectionType,
    pub clip_start: f64,
    pub clip_end: f64,
    #[serde(default)]
    pub is_active: bool,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub depth_of_field: Option<DepthOfField>,
    /// Constraints currently on the camera object.
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub constraints: Vec<String>,
}

/// `camera.list` filters.
#[derive(Debug, Clone, Default, Serialize, Deserialize, JsonSchema)]
pub struct ListCameras {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub name_contains: Option<String>,
    #[serde(default, flatten)]
    pub page: Page,
}

impl Validate for ListCameras {
    fn validate(&self) -> Result<()> {
        self.page.validate()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn fov_and_focal_length_agree() {
        let sensor = 36.0;
        let lens = Lens::Millimetres(50.0);
        let fov = lens.fov_radians(sensor);
        let back = Lens::FovDegrees(fov.to_degrees()).focal_length(sensor);
        assert!((back - 50.0).abs() < 1e-9, "round trip gave {back}");
    }

    #[test]
    fn absurd_lenses_are_rejected() {
        assert!(Lens::Millimetres(0.0).validate().is_err());
        assert!(Lens::FovDegrees(180.0).validate().is_err());
        assert!(Lens::FovDegrees(60.0).validate().is_ok());
    }

    #[test]
    fn clip_planes_must_be_ordered() {
        let settings = CameraSettings {
            clip_start: Some(10.0),
            clip_end: Some(1.0),
            ..Default::default()
        };
        assert!(settings.validate().is_err());
    }

    #[test]
    fn focus_object_and_distance_conflict() {
        let dof = DepthOfField {
            focus_object: Some(ObjectRef::name("Cube")),
            focus_distance: Some(4.0),
            ..Default::default()
        };
        assert!(dof.validate().is_err());
    }

    #[test]
    fn auto_frame_rejects_a_zero_direction() {
        let params = AutoFrame {
            camera: None,
            objects: vec![],
            padding: 0.1,
            direction: Some(Vec3::ZERO),
            keep_position: false,
            aim: true,
            focus: false,
        };
        assert!(params.validate().is_err());
    }

    #[test]
    fn blade_counts_follow_blender() {
        let ok = DepthOfField {
            blades: Some(0),
            ..Default::default()
        };
        assert!(ok.validate().is_ok());
        let ok = DepthOfField {
            blades: Some(6),
            ..Default::default()
        };
        assert!(ok.validate().is_ok());
        let bad = DepthOfField {
            blades: Some(2),
            ..Default::default()
        };
        assert!(bad.validate().is_err());
    }
}

/// `camera.get` / `camera.delete` / `camera.set_active` / `camera.clear_tracking`
#[derive(Debug, Clone, Default, Serialize, Deserialize, JsonSchema)]
pub struct CameraRefParams {
    /// Camera object. Omit to use the scene's active camera.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub camera: Option<ObjectRef>,
}

/// `camera.look_at`
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct CameraLookAt {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub camera: Option<ObjectRef>,
    /// Explicit world-space aim point.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub point: Option<Vec3>,
    /// Aim at an object's bounding-box centre.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub target: Option<ObjectRef>,
}

/// `camera.depth_of_field.update`
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct UpdateDepthOfField {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub camera: Option<ObjectRef>,
    #[serde(flatten)]
    pub depth_of_field: DepthOfField,
}

impl Validate for CameraRefParams {}

impl Validate for CameraLookAt {
    fn validate(&self) -> Result<()> {
        match (self.point, &self.target) {
            (None, None) => Err(BlenderError::invalid_argument(
                "Provide `point` or `target`.",
            )),
            (Some(_), Some(_)) => Err(BlenderError::invalid_argument(
                "Provide `point` or `target`, not both.",
            )),
            _ => self.point.check_finite("point"),
        }
    }
}

impl Validate for UpdateDepthOfField {
    fn validate(&self) -> Result<()> {
        self.depth_of_field.validate()
    }
}
